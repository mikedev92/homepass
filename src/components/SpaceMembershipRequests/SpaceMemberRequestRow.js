// @flow
import React from 'react';
import { Right, Button, ListItem, Left, Body, Thumbnail, Text, Row } from 'native-base';
import moment from 'moment';
import { variables } from '../../theme';

type Props = {
  spaceMemberRequest?: SpaceMemberRequest | null,
  last: boolean,
  approveSpaceMemberRequest: string => () => void,
  rejectSpaceMemberRequest: string => () => void,
};

function SpaceMemberRequestRow({
  spaceMemberRequest,
  last,
  approveSpaceMemberRequest,
  rejectSpaceMemberRequest,
}: Props) {
  if (!spaceMemberRequest || !spaceMemberRequest.user) {
    return null;
  }

  const spaceMemberRequestId = spaceMemberRequest.spaceMemberRequestId;
  if (typeof spaceMemberRequestId !== 'string') {
    return null;
  }

  return (
    <ListItem key={spaceMemberRequest.spaceMemberRequestId} thumbnail last={last}>
      {spaceMemberRequest.user.imageUrl &&
        <Left><Thumbnail source={{ uri: spaceMemberRequest.user.imageUrl }} /></Left>}
      <Body>
        <Text>{spaceMemberRequest.user.name}</Text>
        <Text note>{moment(spaceMemberRequest.createdAt).fromNow()}</Text>
        <Row style={{ marginTop: variables.contentPadding }}>
          <Button
            onPress={approveSpaceMemberRequest(spaceMemberRequestId)}
            style={{ marginRight: variables.contentPadding }}
            small
          >
            <Text>Approve</Text>
          </Button>
          <Button onPress={rejectSpaceMemberRequest(spaceMemberRequestId)} small bordered>
            <Text>Decline</Text>
          </Button>
        </Row>
      </Body>
      <Right />
    </ListItem>
  );
}

export default SpaceMemberRequestRow;
