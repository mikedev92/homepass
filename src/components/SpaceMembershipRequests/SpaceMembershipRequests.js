// @flow
import React from 'react';
import {
  Right,
  Button,
  List,
  Container,
  Header,
  Left,
  Icon,
  Content,
  Body,
  Title,
  ListItem,
  Text,
} from 'native-base';
import { compose } from 'recompose';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import { get } from 'lodash';
import SpaceMemberRequestRow from './SpaceMemberRequestRow';

type Props = {
  data: ApolloData<Space>,
  navigation: Navigation,
  approveSpaceMemberRequest: string => () => void,
  rejectSpaceMemberRequest: string => () => void,
};

function SpaceMembershipRequests({
  navigation,
  data,
  approveSpaceMemberRequest,
  rejectSpaceMemberRequest,
}: Props) {
  const space = data.node;
  if (!space) {
    return null;
  }
  const memberRequests: SpaceMemberRequestEdge[] = get(space, 'memberRequests.edges') || [];
  return (
    <Container backgroundDark>
      <Header bordered>
        <Left>
          <Button onPress={() => navigation.goBack()} transparent primary>
            <Icon name="arrow-back" />
          </Button>
        </Left>
        <Body style={{ flex: 3 }}><Title>Membership Requests</Title></Body>
        <Right />
      </Header>
      <Content
        contentContainerStyle={
          memberRequests.length === 0 ? { flex: 1, alignItems: 'center', justifyContent: 'center' } : {}
        }
        padder={memberRequests.length === 0}
      >
        {memberRequests.length > 0 && <ListItem itemHeader />}
        {memberRequests.length > 0 &&
          <List style={{ flex: 1 }}>
            {memberRequests.map(
              ({ node }, index) =>
                node &&
                <SpaceMemberRequestRow
                  key={node.id}
                  approveSpaceMemberRequest={approveSpaceMemberRequest}
                  rejectSpaceMemberRequest={rejectSpaceMemberRequest}
                  last={index === memberRequests.length - 1}
                  spaceMemberRequest={node}
                />,
            )}
          </List>}
        {memberRequests.length === 0 &&
          <Text caption>There are no membership requests for this space</Text>}
      </Content>
    </Container>
  );
}

const SpaceMembershipRequestsFragment = gql`
  fragment SpaceMembershipRequestsFragment on Space {
    id
    memberRequests {
      edges {
        node {
          id
          spaceMemberRequestId
          createdAt
          user {
            name
            imageUrl
          }
        }
      }
    }
  }
`;
const SpaceMembershipRequestsQuery = gql`
  query SpaceMembershipRequests($id: ID!) {
    node(id: $id) {
      id
      ...SpaceMembershipRequestsFragment
    }
  }
  ${SpaceMembershipRequestsFragment}
`;

const ApproveSpaceMemberRequestMutation = gql`
  mutation($input: ApproveSpaceMemberRequestInput!) {
    approveSpaceMemberRequest(input: $input) {
      space {
        ...SpaceMembershipRequestsFragment
      }
    }
  }
  ${SpaceMembershipRequestsFragment}
`;

const RejectSpaceMemberRequestMutation = gql`
  mutation($input: RejectSpaceMemberRequestInput!) {
    rejectSpaceMemberRequest(input: $input) {
      space {
        ...SpaceMembershipRequestsFragment
      }
    }
  }
  ${SpaceMembershipRequestsFragment}
`;

export default compose(
  graphql(SpaceMembershipRequestsQuery, {
    options: props => ({ variables: { id: props.navigation.state.params.id } }),
  }),
  graphql(ApproveSpaceMemberRequestMutation, {
    props: ({ mutate }) => ({
      approveSpaceMemberRequest: spaceMemberRequestId => () =>
        mutate({
          variables: {
            input: { spaceMemberRequestId },
          },
        }),
    }),
  }),
  graphql(RejectSpaceMemberRequestMutation, {
    props: ({ mutate }) => ({
      rejectSpaceMemberRequest: spaceMemberRequestId => () =>
        mutate({
          variables: {
            input: { spaceMemberRequestId },
          },
        }),
    }),
  }),
)(SpaceMembershipRequests);
