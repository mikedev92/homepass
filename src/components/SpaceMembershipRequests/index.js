import SpaceMembershipRequests from './SpaceMembershipRequests';

export default SpaceMembershipRequests;
