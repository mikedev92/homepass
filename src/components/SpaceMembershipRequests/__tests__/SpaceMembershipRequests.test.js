// @flow
import React from 'react';
import { shallow } from 'enzyme';
import SpaceMembershipRequests from '../SpaceMembershipRequests';

const testProps = [
  {
    data: {
      node: {
        memberRequests: {
          edges: [{}],
        },
      },
    },
  },
  {
    data: {
      node: {
        memberRequests: {
          edges: [],
        },
      },
    },
  },
  {
    data: {
      node: {
        memberRequests: {
          edges: [{}, {}, {}],
        },
      },
    },
  },
];
describe('<SpaceMembershipRequests />', () => {
  it('matches snapshots', () => {
    testProps.forEach((props) => {
      const baseProps = {
        approveSpaceMemberRequest: jest.fn(),
        rejectSpaceMemberRequest: jest.fn(),
      };
      const wrapper = shallow(<SpaceMembershipRequests {...baseProps} {...props} />);
      expect(wrapper).toMatchSnapshot();
    });
  });
});
