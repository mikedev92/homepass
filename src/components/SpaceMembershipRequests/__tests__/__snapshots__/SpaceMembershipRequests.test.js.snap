// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpaceMembershipRequests /> matches snapshots 1`] = `
<Styled(Container)
  backgroundDark={true}
>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <Styled(Title)>
        Membership Requests
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right) />
  </Styled(Header)>
  <Styled(Content)
    contentContainerStyle={Object {}}
    padder={false}
  >
    <Styled(ListItem)
      itemHeader={true}
    />
    <Styled(List)
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<SpaceMembershipRequests /> matches snapshots 2`] = `
<Styled(Container)
  backgroundDark={true}
>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <Styled(Title)>
        Membership Requests
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right) />
  </Styled(Header)>
  <Styled(Content)
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      }
    }
    padder={true}
  >
    <Styled(Text)
      caption={true}
    >
      There are no membership requests for this space
    </Styled(Text)>
  </Styled(Content)>
</Styled(Container)>
`;

exports[`<SpaceMembershipRequests /> matches snapshots 3`] = `
<Styled(Container)
  backgroundDark={true}
>
  <Styled(Header)
    bordered={true}
  >
    <Styled(Left)>
      <Styled(Button)
        onPress={[Function]}
        primary={true}
        transparent={true}
      >
        <Styled(Icon)
          name="arrow-back"
        />
      </Styled(Button)>
    </Styled(Left)>
    <Styled(Body)
      style={
        Object {
          "flex": 3,
        }
      }
    >
      <Styled(Title)>
        Membership Requests
      </Styled(Title)>
    </Styled(Body)>
    <Styled(Right) />
  </Styled(Header)>
  <Styled(Content)
    contentContainerStyle={Object {}}
    padder={false}
  >
    <Styled(ListItem)
      itemHeader={true}
    />
    <Styled(List)
      style={
        Object {
          "flex": 1,
        }
      }
    />
  </Styled(Content)>
</Styled(Container)>
`;
